### ISSUE TITLE
sparse_categorical_crossentropy is slow when training on GPU

### ISSUE THREAD

**Issue by hr0nix** [fchollet/keras#2732]
Hello!

Currently training with sparse_categorical_crossentropy works as follows: it takes a sparse representation of the target, converts it into dense one-hot representation on CPU, copies the resulting dense matrix on GPU and then calls categorical_crossentropy with dense inputs. When training a generative text model with large output vocabulary size (100k+), the time needed to copy a target matrix to GPU can be quite significant (up to 20% in some of my experiments). Given that categorical_crossentropy in theano can take sparse inputs, can we get rid of this inefficiency?

**Comment by lemuriandezapada** [comment]
Does it?
It's all done in keras/theano/tensorflow primitives so I'm quite sure the expansion gets done on the gpu side. It feels faster on my machine. But maybe I'm missing something?

**Comment by hr0nix** [comment]
When I profile a model that uses sparse_categorical_crossentropy loss, I see that AdvancedIncSubtensor, Alloc and GpuFromHost ops take a significant fraction of training time, and these are the only ops in the whole profile that execute on CPU.

**Comment by nouiz** [comment]
We are in the process of implementing AdvancedSubtensor and
AdvancedIncSubtensor on the GPU. Until then you can convert it manually in
your graph to advanced indexing case that we support on the GPU. We support:

a_tensor[a_vector_of_index]

You can convert all advanced indexing to that case. So it is sure this
change can be done.

I don't know when we will finish the new implementation on the GPU.

**Comment by hr0nix** [comment]
That's very good news, but my point is that this problem can probably be fixed by switching to sparse version of theano's categorical_crossentropy, without using AdvancedIncSubtensor at all.

**Comment by cchhdd** [comment]
@hr0nix could you tell me how to use sparse_categorical_crossentropy in keras?indexerror always happens when sparse_categorical_crossentropy in model.fit.  hoping for your help
